[Merge] implement ExecutionPayload timestamp validation in BlockValidator

tbenr opened the pull request:
## PR Description

implements the validation in `BlockValidator` and migrate `BlockValidatorTest` to use `StorageSystem` and `ChainUpdater`

## Fixed Issue(s)
fixes #4669

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

tbenr committed:
implement ExecutionPayload timestamp validation in BlockValidator

tbenr committed:
cleanup

tbenr commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockValidator.java:
```diff
@@ -79,6 +81,25 @@ public BlockValidator(final Spec spec, RecentChainData recentChainData) {
+      if (maybeState.isEmpty()) {
+        LOG.trace(
+            "No state is available to compute timestamp block slot. Block will be saved for future processing");
+        return completedFuture(InternalValidationResult.SAVE_FOR_FUTURE);
```
not sure if this actually makes sense at this stage

tbenr committed:
spotlessly

ajsutton commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/BlockValidatorTest.java:
```diff
@@ -210,4 +243,51 @@ void shouldReturnInvalidForBlockThatDoesNotDescendFromFinalizedCheckpoint() {
+
+  @TestTemplate
+  void shouldReturnAcceptOnCorrectExecutionPayloadTimestamp(SpecContext specContext) {
+    onlyMerge(specContext);
```
nit: Would it make more sense to make this `specContext.onlyMerge()`?  Or probably more correct here would be `specContext.assumeMergeEnabled()` (or `assumeMergeActive`) so that it also runs on forks after the merge.

ajsutton commented on ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/BlockValidatorTest.java:
```diff
@@ -210,4 +243,51 @@ void shouldReturnInvalidForBlockThatDoesNotDescendFromFinalizedCheckpoint() {
+
+    doReturn(Optional.of(specContext.getDataStructureUtil().randomExecutionPayload()))
+        .when(beaconBlockBody)
+        .getOptionalExecutionPayload();
```
Using spies to make this work feels pretty ugly. We generally shouldn't need to mock/spy data structure objects like blocks.
I think we'd ideally use the `generateBlockAtSlot(nextSlot, <BlockOptions>)` method and add the ability to specify a specific `ExecutionPayload` for the created block. Might need to also specify the resulting state root since you won't actually be able to apply the block.

I know it's a bunch more work here but it gives us the test utilities we need to do more tests like this which I think is a pretty big win.

tbenr replied:
I spent some time on that (i implemented the setExecutionPayload) but block processor kicks in even if you are just generating a new block without applying it ant it gets rejected.. so appeared to me that we need a way to skip validations to be able to build "wrong" chains.

tbenr replied:
I updated the test. Have a look if the approach makes sense to you.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockValidator.java:
```diff
@@ -79,6 +81,25 @@ public BlockValidator(final Spec spec, RecentChainData recentChainData) {
 
+    Optional<ExecutionPayload> maybeExecutionPayload =
+        block.getMessage().getBody().getOptionalExecutionPayload();
+    Optional<BeaconState> maybeState = recentChainData.getBestState();
```
We can't use `recentChainData.getBestState()` here because this block might not be on the same fork as our best state.  Further down (line 120) or so we get a state based on the parent block which is what we need to use to check if the merge has been completed and a non-default payload is required.

ajsutton replied:
Given we need a state to verify the timestamp anyway, I'd just push the whole timestamp check down to where we have the relevant state and we can then implement it exactly as the spec does - if mergeComplete(state) then require timestamp == computeTimeAtSlot(..).  No need to explicitly check if the execution payload is default or not because if it is then the timestamp won't match.

tbenr replied:
> We can't use recentChainData.getBestState() here because this block might not be on the same fork as our best state

i was thinking that it was acceptable since under the hood it is only getting the genesis time

> No need to explicitly check if the execution payload is default or not because if it is then the timestamp won't match.

right, also because i suspect isDefault isn't cheep as it may seems, but in isMergeCompleted there is a isDefault check anyway.

ajsutton replied:
True, we would get away with that but we've had a lot of problems in the past because we wound up using the wrong state for things.   We could though just use `recentChainData.getGenesisTime` and have a variant of `computeTimeAtSlot` which takes the genesis time instead of the state (which is often useful anyway). Then we're being explicit about what we're actually using and there won't be any confusion.  We'd still need to check if the payload is non-default if the pre-state had completed the merge though so it's probably more efficient to just do the one check once we have the state.

ajsutton replied:
And yes, that is different to what I was thinking when we were first talking about this. Sorry.

tbenr committed:
implements test block production without state transition, to be able to generate invalid blocks.
apply suggestions in timestamp validation
apply suggestions in SpecContext assumptions helpers
improve a ValidatorDataProvider test

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockValidator.java:
```diff
@@ -101,19 +102,36 @@ public BlockValidator(final Spec spec, RecentChainData recentChainData) {
+                                  .getMessage()
+                                  .getBody()
+                                  .getOptionalExecutionPayload()
+                                  .orElseThrow();
```
nit: I think this is right because to parse the block we'd have to get a payload and we're just checking if it's the default payload or not.  But I think it would be clearer if instead of `.orElseThrow` we just returned `reject("Missing execution payload")` if it's not present.

tbenr replied:
yeah, will never happen but going to change it

ajsutton reviewed (approved):
LGTM.

tbenr committed:
fixes schema

tbenr committed:
Merge branch 'master' into gossip_block_validation_conditions
